File: 01/01.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

func necessaryFuelFor(mass int) int {
	fuel := mass/3 - 2

	if fuel <= 0 {
		return 0
	}

	return fuel + necessaryFuelFor(fuel)
}

func naiveFuelFor(mass int) int {
	return mass/3 - 2
}

func partOneWithStrings(fuelCalculator func(int) int) int {
	input, _ := ioutil.ReadFile("input")
	total := 0
	for _, word := range strings.Split(string(input), "\n") {
		mass, _ := strconv.Atoi(word)
		total += fuelCalculator(mass)
	}
	return total
}

func partOneWithScanner(fuelCalculator func(int) int) int {
	file, _ := os.Open("input")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanWords)

	total := 0
	for scanner.Scan() {
		mass, err := strconv.Atoi(scanner.Text())
		if err != nil {
			panic(err)
		}

		total += fuelCalculator(mass)
	}

	return total
}

func main() {
	fmt.Println("part one (strings):", partOneWithStrings(naiveFuelFor))
	fmt.Println("part one (scanner):", partOneWithScanner(naiveFuelFor))
	fmt.Println("part two (strings):", partOneWithStrings(necessaryFuelFor))
	fmt.Println("part two (scanner):", partOneWithScanner(necessaryFuelFor))
}

File: 01/01.rb
def necessary_fuel_for(mass)
  fuel = mass / 3 - 2

  if fuel <= 0
    0
  else
    fuel + necessary_fuel_for(fuel)
  end
end

def mass_without_the_fuel
  puts File.read('input')
    .lines
    .map { |mass| mass.to_i / 3 - 2 }
    .sum(0)
end

def mass_accounting_for_the_fuel
  puts File.read('input')
    .lines
    .map { |mass| necessary_fuel_for mass.to_i }
    .sum(0)
end

mass_accounting_for_the_fuel
